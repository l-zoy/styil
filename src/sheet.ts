/**
 * https://github.com/emotion-js/emotion/tree/main/packages/sheet
 */

import { Rules, StyleSheetOptions } from './types'

export class StyleSheet {
  /**
   * used to record whether another rule is inserted before
   */
  private alreadyInsertedOrderInsensitiveRule = false

  /**
   * used to record the number of inserts
   */
  private insertIndex = 0

  /**
   * use insertRule, the production default `true`, and the development environment default to `appendChild`
   * @def
   */
  speedy: boolean

  /**
   * insert a list of style tags
   */
  tags: HTMLStyleElement[] = []

  /**
   * Using Node instead of HTMLElement since container may be a ShadowRoot
   */
  container: Node

  /**
   * random value
   * https://web.dev/csp/
   */
  nonce?: string

  /**
   * this will be set as the value of the style tag attribute that get inserted.
   */
  key: string

  /**
   * ssr data
   */
  ssrData = ''

  constructor(options: StyleSheetOptions) {
    this.speedy = options.speedy
    this.nonce = options.nonce
    this.key = options.key
    this.container = options.container
  }

  insertTag(tag: HTMLStyleElement) {
    let before: Node | null = null
    if (this.tags.length !== 0) {
      before = this.tags[this.tags.length - 1].nextSibling
    }
    this.container.insertBefore(tag, before)

    this.tags.push(tag)
  }

  insertStyle({ ruleCode, segmentRuleCode }: Rules) {
    if (this.container) {
      if (this.speedy) {
        for (let index = 0; index < segmentRuleCode.length; index++) {
          this.insert(segmentRuleCode[index])
        }
      } else {
        this.insert(ruleCode)
      }
    } else {
      this.ssrData += ruleCode
    }
  }

  insert(rule: string) {
    // the max length is how many rules we have per style tag, it's 65000 in speedy mode
    // it's 1 in dev because we insert source maps that map a single rule to a location
    // and you can only have one source map per style tag
    if (this.insertIndex % (this.speedy ? 65000 : 1) === 0) {
      const tag = document.createElement('style')

      tag.setAttribute('data-styil', this.key)
      if (this.nonce !== undefined) {
        tag.setAttribute('nonce', this.nonce)
      }
      tag.appendChild(document.createTextNode(''))

      this.insertTag(tag)
    }
    const tag = this.tags[this.tags.length - 1]

    if (process.env.NODE_ENV !== 'production') {
      // is `@import`
      const isImportRule = rule.charCodeAt(0) === 64 && rule.charCodeAt(1) === 105

      if (isImportRule && this.alreadyInsertedOrderInsensitiveRule) {
        // this would only cause problem in speedy mode
        // but we don't want enabling speedy to affect the observable behavior
        // so we report this error at all times
        console.error(
          [
            `You're attempting to insert the following rule:\n${rule}\n\n`,
            '`@import` rules must be before all other types of rules in a stylesheet but other rules have already been inserted.',
            'Please ensure that `@import` rules are before all other rules.'
          ].join('')
        )
      }

      this.alreadyInsertedOrderInsensitiveRule =
        this.alreadyInsertedOrderInsensitiveRule || !isImportRule
    }

    if (this.speedy) {
      try {
        // this is the ultrafast version, works across browsers
        // the big drawback is that the css won't be editable in devtools
        // here tag.sheet is required unless manually modified
        tag.sheet!.insertRule(rule, tag.sheet!.cssRules.length)
      } catch (error) {
        if (
          process.env.NODE_ENV !== 'production' &&
          !/:(-moz-placeholder|-moz-focus-inner|-moz-focusring|-ms-input-placeholder|-moz-read-write|-moz-read-only|-ms-clear){/.test(
            rule
          )
        ) {
          console.error(
            `There was a problem inserting the following rule: "${rule}"`,
            error?.message
          )
        }
      }
    } else {
      tag.appendChild(document.createTextNode(rule))
    }

    this.insertIndex++
  }

  flush() {
    this.tags.forEach((tag) => tag.parentNode && tag.parentNode.removeChild(tag))
    this.ssrData = ''
    this.tags = []
    this.insertIndex = 0

    if (process.env.NODE_ENV !== 'production') {
      this.alreadyInsertedOrderInsensitiveRule = false
    }
  }
}
