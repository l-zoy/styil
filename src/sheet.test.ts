import { StyleSheet } from './sheet'

const rule = 'html { color: hotpink; }'
const rule2 = '* { box-sizing: border-box; }'

const defaultOptions = {
  key: '',
  container: document.head
}

console.error = jest.fn()

afterEach(() => {
  jest.clearAllMocks()
})

beforeEach(() => {
  document.head.innerHTML = ''
  document.body.innerHTML = ''
})

describe('StyleSheet', () => {
  it('should be speedy by default in production', () => {
    const sheet = new StyleSheet({ ...defaultOptions, speedy: true })
    expect(sheet.speedy).toBe(true)
    process.env.NODE_ENV = 'test'
  })

  it('should not be speedy in a non-production environment by default', () => {
    const sheet = new StyleSheet(defaultOptions)
    expect(sheet.speedy).toBe(undefined)
  })

  it('should remove its style elements from the document when flushed', () => {
    const sheet = new StyleSheet(defaultOptions)
    sheet.insert(rule)
    expect(document.documentElement).toMatchSnapshot()
    sheet.flush()
    expect(document.documentElement).toMatchSnapshot()
  })

  it('should set the data-styls attribute to the key option', () => {
    const key = 'some-key'
    const sheet = new StyleSheet({ ...defaultOptions, key })
    sheet.insert(rule)
    expect(document.documentElement).toMatchSnapshot()
    expect(document.querySelector('[data-styls]')?.getAttribute('data-styls')).toBe(key)
    sheet.flush()
  })

  it('should insert a rule into the DOM when not in speedy', () => {
    const sheet = new StyleSheet(defaultOptions)
    sheet.insert(rule)
    expect(document.documentElement).toMatchSnapshot()
    sheet.flush()
  })

  it('should insert a rule with insertRule when in speedy', () => {
    const sheet = new StyleSheet({ ...defaultOptions, speedy: true })
    sheet.insert(rule)
    expect(document.documentElement).toMatchSnapshot()
    expect(sheet.tags).toHaveLength(1)

    expect(sheet.tags[0].sheet?.cssRules).toMatchSnapshot()
    sheet.flush()
  })

  it('should throw when inserting a bad rule in speedy mode', () => {
    const sheet = new StyleSheet({ ...defaultOptions, speedy: true })
    sheet.insert('.asdfasdf4###112121211{')
    expect(console.error).toHaveBeenCalledTimes(1)

    sheet.flush()
  })

  it('should set the nonce option as an attribute to style elements', () => {
    const nonce = 'some-nonce'
    const sheet = new StyleSheet({ ...defaultOptions, nonce })
    sheet.insert(rule)
    expect(sheet.tags[0]).toBe(document.querySelector('[data-styls]'))
    expect(sheet.tags).toHaveLength(1)
    expect(sheet.tags[0].getAttribute('nonce')).toBe(nonce)
    sheet.flush()
  })

  it("should use the container option instead of document.head to insert style elements into if it's passed", () => {
    const container = document.createElement('div')
    document.body
    document.body.appendChild(container)
    const sheet = new StyleSheet({ ...defaultOptions, container })
    expect(sheet.container).toBe(container)
    sheet.insert(rule)
    expect(document.documentElement).toMatchSnapshot()
    expect(sheet.tags).toHaveLength(1)
    expect(sheet.tags[0].parentNode).toBe(container)
    sheet.flush()
  })

  it('should work with a ShadowRoot container', () => {
    const div = document.createElement('div')
    document.body
    document.body.appendChild(div)
    const container = div.attachShadow({ mode: 'open' })
    const sheet = new StyleSheet({ ...defaultOptions, container })
    expect(sheet.container).toBe(container)
    sheet.insert(rule)
    expect(document.documentElement).toMatchSnapshot()
    // The shadowRoot is not serialized in the snapshot, so we need to take a
    // separate snapshot of the shadowRoot's children.
    expect(container.children).toMatchSnapshot()
    expect(sheet.tags).toHaveLength(1)
    expect(sheet.tags[0].parentNode).toBe(container)
    sheet.flush()
  })

  it('should accept prepend option', () => {
    const { head } = document
    const otherStyle = document.createElement('style')
    otherStyle.setAttribute('id', 'other')
    head.appendChild(otherStyle)

    const sheet = new StyleSheet({ ...defaultOptions, prepend: true })
    sheet.insert(rule)
    sheet.insert(rule2)
    expect(document.documentElement).toMatchSnapshot()

    sheet.flush()
  })

  it('should accept insertionPoint option', () => {
    const { head } = document

    head.innerHTML = `
      <style id="first"></style>
      <style id="last"></style>
    `

    // the sheet should be inserted between the first and last style nodes
    const sheet = new StyleSheet({
      ...defaultOptions
    })
    sheet.insert(rule)
    sheet.insert(rule2)
    expect(document.documentElement).toMatchSnapshot()

    sheet.flush()
  })

  it('should work if insertionPoint is last element', () => {
    const { head } = document
    const lastStyle = document.createElement('style')
    lastStyle.setAttribute('id', 'last')
    head.appendChild(lastStyle)

    // the sheet should be inserted after the first node
    const sheet = new StyleSheet({
      ...defaultOptions
    })
    sheet.insert(rule)
    sheet.insert(rule2)
    expect(document.documentElement).toMatchSnapshot()

    sheet.flush()
  })

  it('should be able to hydrate styles', () => {
    const fooStyle = document.createElement('style')
    fooStyle.textContent = '.foo { color: hotpink; }'
    const barStyle = document.createElement('style')
    barStyle.textContent = '.bar { background-color: green; }'
    const { body } = document
    body.appendChild(fooStyle)
    body.appendChild(barStyle)

    const sheet = new StyleSheet(defaultOptions)
    expect(document.documentElement).toMatchSnapshot()

    expect(document.documentElement).toMatchSnapshot()

    sheet.flush()
  })

  it('should flush hydrated styles', () => {
    const fooStyle = document.createElement('style')
    fooStyle.textContent = '.foo { color: hotpink; }'
    const barStyle = document.createElement('style')
    barStyle.textContent = '.bar { background-color: green; }'
    const { body } = document
    body.appendChild(fooStyle)
    body.appendChild(barStyle)

    const sheet = new StyleSheet(defaultOptions)

    sheet.insert(rule)
    sheet.insert(rule2)
    expect(document.documentElement).toMatchSnapshot()

    sheet.flush()

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should correctly position hydrated styles when used with `prepend` option', () => {
    const { head } = document
    const otherStyle = document.createElement('style')
    otherStyle.setAttribute('id', 'other')
    head.appendChild(otherStyle)

    const fooStyle = document.createElement('style')
    fooStyle.textContent = '.foo { color: hotpink; }'
    const barStyle = document.createElement('style')
    barStyle.textContent = '.bar { background-color: green; }'
    const { body } = document
    body.appendChild(fooStyle)
    body.appendChild(barStyle)

    const sheet = new StyleSheet({
      ...defaultOptions,
      prepend: true
    })

    expect(document.documentElement).toMatchSnapshot()

    sheet.flush()
  })

  it('should not crash when flushing when styles are already detached', () => {
    const { head } = document

    const sheet = new StyleSheet(defaultOptions)

    sheet.insert(rule)

    head.innerHTML = ''

    expect(() => sheet.flush()).not.toThrowError()
  })

  it('`@import` rules must be before all other types of rules in a stylesheet but other rules have already been inserted.', () => {
    const sheet = new StyleSheet(defaultOptions)

    sheet.insert(rule)
    sheet.insert('@import url(http://google.com)')

    expect(console.error).toHaveBeenCalledWith(
      [
        `You're attempting to insert the following rule:\n@import url(http://google.com)\n\n`,
        '`@import` rules must be before all other types of rules in a stylesheet but other rules have already been inserted.',
        'Please ensure that `@import` rules are before all other rules.'
      ].join('')
    )
  })
})
