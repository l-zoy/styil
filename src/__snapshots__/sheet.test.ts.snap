// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet should accept insertionPoint option 1`] = `
<html>
  <head>
    
      
    <style
      id="first"
    />
    
      
    <style
      id="last"
    />
    
    
    <style
      data-styils=""
    >
      
      html { color: hotpink; }
    </style>
    <style
      data-styils=""
    >
      
      * { box-sizing: border-box; }
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should accept prepend option 1`] = `
<html>
  <head>
    <style
      id="other"
    />
    <style
      data-styils=""
    >
      
      html { color: hotpink; }
    </style>
    <style
      data-styils=""
    >
      
      * { box-sizing: border-box; }
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should be able to hydrate styles 1`] = `
<html>
  <head />
  <body>
    <style>
      .foo { color: hotpink; }
    </style>
    <style>
      .bar { background-color: green; }
    </style>
  </body>
</html>
`;

exports[`StyleSheet should be able to hydrate styles 2`] = `
<html>
  <head />
  <body>
    <style>
      .foo { color: hotpink; }
    </style>
    <style>
      .bar { background-color: green; }
    </style>
  </body>
</html>
`;

exports[`StyleSheet should correctly position hydrated styles when used with \`prepend\` option 1`] = `
<html>
  <head>
    <style
      id="other"
    />
  </head>
  <body>
    <style>
      .foo { color: hotpink; }
    </style>
    <style>
      .bar { background-color: green; }
    </style>
  </body>
</html>
`;

exports[`StyleSheet should flush hydrated styles 1`] = `
<html>
  <head>
    <style
      data-styils=""
    >
      
      html { color: hotpink; }
    </style>
    <style
      data-styils=""
    >
      
      * { box-sizing: border-box; }
    </style>
  </head>
  <body>
    <style>
      .foo { color: hotpink; }
    </style>
    <style>
      .bar { background-color: green; }
    </style>
  </body>
</html>
`;

exports[`StyleSheet should flush hydrated styles 2`] = `
<html>
  <head />
  <body>
    <style>
      .foo { color: hotpink; }
    </style>
    <style>
      .bar { background-color: green; }
    </style>
  </body>
</html>
`;

exports[`StyleSheet should insert a rule into the DOM when not in speedy 1`] = `
<html>
  <head>
    <style
      data-styils=""
    >
      
      html { color: hotpink; }
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should insert a rule with insertRule when in speedy 1`] = `
<html>
  <head>
    <style
      data-styils=""
    >
      
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should insert a rule with insertRule when in speedy 2`] = `
Array [
  CSSStyleRule {
    "__ends": 24,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "selectorText": "html",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 5,
      "_importants": Object {
        "color": "",
      },
      "color": "hotpink",
      "length": 1,
      "parentRule": [Circular],
    },
  },
]
`;

exports[`StyleSheet should remove its style elements from the document when flushed 1`] = `
<html>
  <head>
    <style
      data-styils=""
    >
      
      html { color: hotpink; }
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should remove its style elements from the document when flushed 2`] = `
<html>
  <head />
  <body />
</html>
`;

exports[`StyleSheet should set the data-styils attribute to the key option 1`] = `
<html>
  <head>
    <style
      data-styils="some-key"
    >
      
      html { color: hotpink; }
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should use the container option instead of document.head to insert style elements into if it's passed 1`] = `
<html>
  <head />
  <body>
    <div>
      <style
        data-styils=""
      >
        
        html { color: hotpink; }
      </style>
    </div>
  </body>
</html>
`;

exports[`StyleSheet should work if insertionPoint is last element 1`] = `
<html>
  <head>
    <style
      id="last"
    />
    <style
      data-styils=""
    >
      
      html { color: hotpink; }
    </style>
    <style
      data-styils=""
    >
      
      * { box-sizing: border-box; }
    </style>
  </head>
  <body />
</html>
`;

exports[`StyleSheet should work with a ShadowRoot container 1`] = `
<html>
  <head />
  <body>
    <div />
  </body>
</html>
`;

exports[`StyleSheet should work with a ShadowRoot container 2`] = `
HTMLCollection [
  <style
    data-styils=""
  >
    
    html { color: hotpink; }
  </style>,
]
`;
